rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1080 @@
+- Formatting non-Nix files in Nixpkgs (that's for [treefmt](https://github.com/numtide/treefmt))
+- Applying the format to other repositories within the NixOS organization containing Nix code.
+  It is up to their respective maintainers to make the transition.
+
+## Goals and approach
+
+There are several goals that the formatting style should match.
+These are inherently in conflict with each other, requiring prioritisation and making trade-offs.
+The resulting choice is always a compromise.
+
+In general, we want the code to be (in no particular order):
+
+- **Short and concise.** Code should not be spread across too many lines, but also without being crammed
+- **Readable.** The output format should reflect the semantic flow of the program.
+  It should be clear where expressions start and end.
+  The amount of information per line should be limited.
+- **Consistent.** Similar syntax constructs should be formatted similarly.
+    - The number of special cases in the formatting rules should be minimized.
+- **Diffable and stable.** Small changes to the code should not result in excessive changes in the output.
+
+The general approach taken here is to liberally expand expressions by default, with the goal of being stable, diffable and consistent.
+Then, special cases with more compact output for the most common patterns are introduced as needed,
+sacrificing those properties in favor of conciseness.
+The idea is that this results in a format that is spread-out by default but compact where it matters.
+
+The interactions between different language features are complex and producing a style that matches the expectations involves a lot of special cases.
+Any attempt at creating an exhaustive rule set would be futile.
+Therefore, the formatting rules are intentionally under-specified, leaving room for the formatter implementation.
+However, the most important or potentially controversial rules are included, as well as some general meta-rules.
+
+When deciding between two *equally good* options, currently prevalent formatting style in Nixpkgs should be followed.
+The emphasis here is on "equally good".
+We should not fear making radical changes to the current style if there are sufficient arguments in favor of it.
+
+*Bad code does not deserve good formatting.*
+
+## Detailed design
+
+### Standard Nix format
+
+The _standard Nix format_ defines the officially recommended way that Nix code should be formatted.
+
+The initial version of the standard Nix format is defined in a section towards the end:
+
+[Initial standard Nix format](#initial-standard-nix-format).
+
+Significant changes to the standard Nix format must go through another RFC.
+
+The latest version of the standard Nix format must be in a file on the main branch of the [official Nix formatter](#official-nix-formatter).
+
+### Nix format team
+
+The _Nix format team_ is established, it has the responsibility to
+- Maintain the [official Nix formatter](#official-nix-formatter)
```

Comment:
We have a team of 5 people from the community that are _willing_ to maintain the formatter, which I think it way better than just hoping somebody will maintain it!

Reply:
I like this crucial step of creating the team. 👍 Thank you so much for tackling this problem.

I still think @hugosenari has an important point here: The nix language formatting specification should be *abstractly* written down so other oss tools (e..g a lsp that warns about certain violations) can implement the same behavior without reverse engineering the 'official tool'.

Reply:
We mainly want to avoid formatting debates in Nixpkgs by only having one way to format Nix code, checked by CI. But it's really hard to write down a _complete_ specification of the formatting, at least from the start. By declaring a formatter as official, the formatter team can make minor adjustments for edge cases and fix bugs without having to go through another RFC every time. Everything not specified in the RFC is implicitly specified by the implementation and can be adjusted as necessary.